### PR TITLE
Add 2d tiff backend

### PR DESCRIPTION
**Pull request by fm3**
requires tiled tifs with resolution pyramid, as produced by
`vips tiffsave source.tif color.tif --tile --pyramid --bigtiff --compression none --tile-width 256 --tile-height 256`
(make sure you have enough space at /tmp or change your TMPDIR env var)
or
`convert source.tif -define tiff:tile-geometry=256x256 'ptif:color.tif'`
(note that imagemagick/convert fails for >4GB images, as it does not support resolution pyramid and bigtiff at the same time)

In `data/datasets.json`, add
```
    "tiff": {
        "sample_organization": {
            "my_tiff_dataset": {
                "scale": [1,1]
            }
        }
    }
```
and in `data/binary/sample_organization/my_tiff_dataset`, place the generated `color.tif`

TODO:
 - [x] read tiles from tifs using numpy.mmap
 - [x] handle resolution pyramid
 - [x] add caching, add clear-cache-route
 - [x] how to discern layer categories? segmentation layers? → layer_name.startswith("segmentation")
 - [x] handle multiple layers
 - [x] handle dtypes other than uint8
 - [x] handle tif filenames other than "0.tif"
 - [x] add readable error messages for incompatible tifs
 - [x] test with bigtiff (>4GB)

### CODE REVIEW

File: wkconnect/backends/tiff/models.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    ]:
+        filepath = self.layer_filepath(layer_name)
+        with TiffFile(str(filepath)) as tif:
+            mags = [Vec3D(2 ** mag, 2 ** mag, 1) for mag in range(len(tif.pages))]
+            assert (
+                len(mags) > 0
+            ), f"No magnifications found. Empty tiff at {str(filepath)}?"
+            tags = tif.pages[0].tags
+            assert (
+                322 in tags
```
Comment (jstriebel) [inline wkconnect/backends/tiff/models.py:110]: Please extract into a semantic dict

File: wkconnect/backends/tiff/models.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+    @lru_cache(maxsize=2 ** 9)
+    def read_mmapped(
+        self, layer_name: str, dtype: np.dtype, byte_offset: int, shape: Tuple[int, int]
+    ) -> np.ndarray:
+        image = np.zeros(shape, dtype=dtype)
+        mapped = np.memmap(
+            str(self.layer_filepath(layer_name)), dtype, "r", byte_offset, shape, "C"
+        )
+        image[:] = mapped
```
Comment (jstriebel) [inline wkconnect/backends/tiff/models.py:166]: Please add a comment about the file handles

File: wkconnect/backends/tiff/models.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        )
+
+    @lru_cache(maxsize=2 ** 12)
+    def read_data(
+        self, layer_name: str, zoom_step: int, wk_offset: Vec3D, shape: Vec3D
+    ) -> Optional[np.ndarray]:
+        mag_factor = 2 ** zoom_step
+        wk_offset_scaled = wk_offset // mag_factor
+
+        tile_coordinate_offset, tile_data = self.read_tile(
```
Comment (jstriebel) [inline wkconnect/backends/tiff/models.py:70]: Please add a ssertion that wk_offset + shape stays within the requested tile.

File: wkconnect/backends/tiff/models.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    ) -> Tuple[
+        List[Vec3D], Tuple[int, int], List[Tuple[int, int]], np.dtype, List[List[int]]
+    ]:
+        filepath = self.layer_filepath(layer_name)
+        with TiffFile(str(filepath)) as tif:
+            mags = [Vec3D(2 ** mag, 2 ** mag, 1) for mag in range(len(tif.pages))]
+            assert (
+                len(mags) > 0
+            ), f"No magnifications found. Empty tiff at {str(filepath)}?"
+            tags = tif.pages[0].tags
```
Comment (jstriebel) [inline wkconnect/backends/tiff/models.py:108]: Probably the sizes need to be asserted for all pages.

**Comment by fm3** [comment]
@jstriebel I implemented your feedback and tested with the large tif (had to introduce the dtype adapter for signed data). could you have another look?

File: wkconnect/backends/tiff/models.py [hunk 5]
```diff
@@ -97,52 +100,84 @@ def clear_cache(self) -> None:
...
+        List[Tuple[int, int]],
+        np.dtype,
+        List[List[int]],
     ]:
+        # Accessing the tiff header tags via their numbers:
+        tag_tile_width = 322
+        tag_tile_height = 323
+        tag_tile_byte_offsets = 324
+        tag_page_width = 256
+        tag_page_height = 257
```
Comment (jstriebel) [inline wkconnect/backends/tiff/models.py:114]: Maybe use capital-letters to indicate they are constants?
```suggestion
        TAG_TILE_WIDTH = 322
        TAG_TILE_HEIGHT = 323
        TAG_TILE_BYTE_OFFSETS = 324
        TAG_PAGE_WIDTH = 256
        TAG_PAGE_HEIGHT = 257
```

Or maybe directly use `TAGS()` from `tifffile`?
See https://github.com/cgohlke/tifffile/blob/master/tifffile/tifffile.py#L9413

**Review by jstriebel — COMMENTED** [review]
Nice, mostly LGTM :+1: Just one more note regarding the tiff-tags.

Could you also add a note about the tiff-format in the README? Maybe also how to convert an image to the correct format.
And would it make sense to include a minimal tiff example directly in the repo? Not sure how small the tiff could possibly be.

**Comment by fm3** [comment]
Done and done :) I don’t think it makes that much sense to include a tiff dataset by default in the repo, as we don’t have a real-world dataset that we can use and I don’t think we should include too much stuff that users have to remove first when setting this up for real. What do you think?

**Review by jstriebel — APPROVED** [review]
Awesome, LGTM :shipit: 

> Done and done :) I don’t think it makes that much sense to include a tiff dataset by default in the repo, as we don’t have a real-world dataset that we can use and I don’t think we should include too much stuff that users have to remove first when setting this up for real. What do you think?

Agreed :+1:

**Pull request merged**